src/Spectre.libGaussianMixtureModelling.Tests/ExpectationMaximizationTest.cpp

```diff
@@ -0,0 +1,287 @@
+    {
+        std::mt19937_64 rngEngine(0);
+
+        ExpectationMaximization<
+            BasicInitializationRef,
+            ExpectationRunnerRef,
+            MaximizationRunnerRef,
+            LogLikelihoodCalculator
+        > em(&mzs[0], &intensities[0], (unsigned)mzs.size(), 
+            rngEngine, (unsigned)gaussianComponents.size());
+
+        GaussianMixtureModel result = em.EstimateGmm();
+        std::vector<double> approximates = GenerateValues(mzs, result.components);
+
+        std::vector<double> errors(intensities.size());
+        for (unsigned i = 0; i < intensities.size(); i++)
+        {
+            errors[i] = pow(intensities[i] - approximates[i], 2) * 10000.0; // scale the errors
+        }
+
+        double averageError = 0.0;
+        for (unsigned i = 0; i < errors.size(); i++)
+        {
+            averageError += errors[i];
+        }
+        averageError /= errors.size();
+        double maxError = *std::max_element(errors.begin(), errors.end());
+        double minError = *std::min_element(errors.begin(), errors.end());
+        printf("Avg error: %f\n", averageError);
+        printf("Max error: %f\n", maxError);
+        printf("Min error: %f\n", minError);
+
+        EXPECT_LE(averageError, 0.1);
+    }
+
+    TEST_F(ExpectationMaximizationTest, test_em_ref_initialization)
+    {
+        std::mt19937_64 rngEngine(0);
+        std::vector<GaussianComponent> components(gaussianComponents.size());
+        BasicInitializationRef initialization(&mzs[0], (unsigned)mzs.size(),
+            components, rngEngine);
+
+        // Check if values of assigned means come from available mz values
+        initialization.AssignRandomMeans();
+
+        for (int i = 0; i < components.size(); i++)
+        {
+            EXPECT_NE(std::find(mzs.begin(), mzs.end(), components[i].mean), mzs.end());
+        }
+
+        // Calculate sample variance
+        initialization.AssignVariances();
+
+        double sampleMean = std::accumulate(mzs.begin(), mzs.end(), 0.0) / (double)mzs.size();
+        double variance = 0.0;
+        for (unsigned i = 0; i < mzs.size(); i++)
+        {
+            variance += pow(mzs[i] - sampleMean, 2);
+        }
+        variance /= (double)mzs.size();
+        double deviation = sqrt(variance);
+
+        // Check if sample variance was properly assigned to all components
+        for (unsigned i = 0; i < components.size(); i++)
+        {
+            EXPECT_EQ(components[i].deviation, deviation);
+        }
+
+        // Check if weights have been assigned uniformly
+        initialization.AssignWeights();
+
+        double weight = 1.0 / (double)components.size();
+        for (unsigned i = 0; i < components.size(); i++)
+        {
+            EXPECT_EQ(components[i].weight, weight);
+        }
+    }
+
+    TEST_F(ExpectationMaximizationTest, test_em_ref_expectation)
+    {
+        DataType** affilationMatrix = new DataType*[mzs.size()];
```

Comment:
I can agree on this so far, but it would be good to at least add some clarification for this, or extract it to function (and comment or something).